torch/csrc/jit/codegen/cuda/lower_sync_information.cpp

```diff
@@ -0,0 +1,418 @@
+              redundant_preds.hasTID()) {
+            redundant_preds.clearAllBID();
+            raw_dims |= redundant_preds;
+            continue;
+          }
+
+          // When the producer axis is a broadcast, it is not really
+          // parallelized unless thread-predicated and concretized
+          if (isParallelTypeThread(producer_ptype) && p_id->isBroadcast() &&
+              (!parallel_bcast_doms.get(producer_ptype) ||
+               !GpuLower::current()
+                    ->concretizedBroadcastDomains()
+                    .isConcretized(p_id))) {
+            continue;
+          }
+
+          // If matching dims and matching parallel types, no comm is necessary.
+          if (producer_ptype == consumer_ptype &&
+              loop_map.areMapped(p_id, c_id)) {
+            continue;
+          }
+
+          // Set parallel dimensions that communication is occuring over.
+          if (isParallelTypeThread(producer_ptype)) {
+            raw_dims.set(producer_ptype);
+          }
+        } // end for ptypes
+
+        if (raw_dims.hasBID()) {
+          TORCH_INTERNAL_ASSERT(
+              producer->getMemoryType() == MemoryType::Global,
+              "Inconsistent parallelization found between TV",
+              producer->name(),
+              " (",
+              producer->toString(),
+              ") and TV",
+              consumer->name(),
+              "(",
+              consumer->toString(),
+              "). Producer is required to be in Global Memory based on parallelization strategy.");
+        } else if (raw_dims.hasTID()) {
+          TORCH_INTERNAL_ASSERT(
+              producer->getMemoryType() == MemoryType::Global ||
+                  producer->getMemoryType() == MemoryType::Shared,
+              "Inconsistent parallelization found between TV",
+              producer->name(),
+              " (",
+              producer->toString(),
+              ") and TV",
+              consumer->name(),
+              "(",
+              consumer->toString(),
+              "). Producer is required to be in Global or Shared Memory based on parallelization strategy.");
+        }
+
+      } // end for consumers
+
+      if (raw_dims.any()) {
+        needs_raw_sync_[producer] = raw_dims;
+
+        bool needs_war = false;
+        // War sync only relevant if producer is in smem
+        if (producer->getMemoryType() == MemoryType::Shared) {
+          for (const auto producer_i : c10::irange(producer->nDims())) {
+            if (producer_i < producer->getComputeAtPosition()) {
+              auto producer_axis = producer->axis(producer_i);
+              auto producer_ptype =
+                  par_map.getConcreteMappedID(producer_axis)->getParallelType();
+              if (!isParallelTypeThread(producer_ptype)) {
+                if (!(producer_axis->isReduction() ||
+                      loop_map.getConcreteMappedID(producer_axis)
+                          ->isBroadcast())) {
```

Comment:
nit: This logic is basically a duplicate of:

```
within_iter_loop_ = within_iter_loop_ || !for_loop->isTrivial();
```
https://github.com/csarofeen/pytorch/pull/1484/files#diff-78b5f3903310443ecb043f667eecd1a1ad60fc478615b8f97e88e3c76a340f4aR244